### DISCUSSION TITLE
Home Assistant Add-on

### DISCUSSION THREAD

**Discussion by Taylorp913** [mrlt8/docker-wyze-bridge#68] [Ideas]
I have been experimenting with making this docker container work on home assistant inside the supervisor with some results. If we added a config.json and setup the container to load environment variables the supervisor way, anyone could import the repo into home assistant. 
Home assistant supervisor can also help optimize the docker image during the build if your willing to track another dockerfile using their base image arguments.

**Comment by mrlt8** [comment]
I added some basic support for Home Assistant in v0.5.15.

Should be able to add the bridge directly in Home Assistant by going to Supervisor > Add-on Store > (three dots on top right) > Repositories, and add:
`https://github.com/mrlt8/docker-wyze-bridge`

Some basic features:
- Will automatically grab the thumbnail from wyze and save to /config/www/cam-name.jpg
  which should be accessible at http://127.0.0.1:8123/local/cam-name.jpg
- 2FA code can be added to /config/mfa_token

**Reply by unrly** [reply]
Been using this for the last day on one camera to test - its fantastic!! For a Docker noob, the integration is so much easier and is an excellent addition to Frigate.

Can any of the whitelisting options be used? I'm about to flash my other cams off the RTSP firmware, excited to move on from that.

edit: Just filtered my unused camera out. Nevermind!

**Reply by mrlt8** [reply]
Awesome! How were you able to get filtering working?

**Reply by unrly** [reply]
In the addon Configuration, click the ellipsis and click "Edit in YAML". I added this to the username/password lines:
`FILTER_NAMES: Grill Cam`
`FILTER_MODE: BLOCK`

**Comment by tbrausch** [comment]
This addon is fantastic!  I now can get rtsp streams of my V3 cameras in home assistant.  Thanks so much for your efforts!  I do have one question.  In the docs it says that camera snapshots should work, but I have not been able to accomplish this whether I define the cameras as ffmpeg or generic.  It also says "The add-on should automatically grab a thumbnail for each of your cameras from wyze and save it to /config/www/cam-name.jpg which should be accessible at http://127.0.0.1:8123/local/cam-name.jpg".  However, these thumbnails are never created.  Do I have a permissions problem?

**Reply by unrly** [reply]
I'm not entirely sure about using a url but I found mine when using the File Editor addon in /config/www/
![image](https://user-images.githubusercontent.com/24898502/131402533-c17e1f18-447b-4a7f-bc62-5d919f8320ed.png)

If you have the Samba addon, you can find them that way as well from another PC on your network.

**Reply by tbrausch** [reply]
Yes, I know where and how to see the directory, but no snapshots are being created.  I have 6 picture entity items on my main screen, one for each camera.  Previously (using Tiny Cam for the V3 cameras), the picture entities displayed am image of each camera.  They were updated automatically every 10 seconds.  I had a button which would change the picture entities to display the cameras in live view.  With this addon, all cameras work perfectly in live view, but there are no snapshots created when live view is disabled and nothing is displayed.  (And hopefully, this is the correct place to ask questions)

**Reply by mrlt8** [reply]
It's supposed to pull the same preview thumbnail that the wyze app uses *if available* and you'd see a message in your log saying something like `☁️ Pulling thumbnail for..`. 
This only happens once at startup to reduce writes to your drive.

If you want an updated image to be displayed, you could try the [ffmpeg platform](https://github.com/mrlt8/docker-wyze-bridge/wiki/Home-Assistant#ffmpeg-camera).

**Comment by tbrausch** [comment]
That's the platform I am using (ffmpeg), but it is not generating still
images.  The video, however, is terrific!  I am getting much higher frame
rates than with Tiny Cam Pro with V3 cameras or RTSP on version 2 cameras.

**Comment by tbrausch** [comment]
Really not sure what is going on, but all day yesterday my picture entity cards would not display the 10 second snapshots.  No changes were made, and now this morning they are working perfectly!  Must be something strange with my setup, but sorry for wasting your time with these questions.  I can't stress enough how fantastic this addon is.  No more waiting for RTSP from Wyze!  In fact, this addon streams V2 cameras much better than RTSP on version 2's.  Thanks so much again for all of your work.

**Comment by esenterre** [comment]
Interesting addon! I have a request, if it's possible at all. Can we set custom ports for the differents streams ? I already have a docker (Frigate) that use port 1935... Can we set it to something else ? With docker I was able to set it to 1936, simply.

**Reply by Taylorp913** [reply]
If you are running this as an Add-on in Home Assistant Supervisor, you can open the add-on in supervisor and change the ports in the Configuration tab.

**Comment by chubRock1** [comment]
Been using the HA add on for a week plus now and am pulling the RTSP streams into Blue Iris from the add on.  I keep having a strange issue, when I first start the add on, all my cameras work fine but after a few hours a few start dropping out of Blue Iris and I start getting "Err no one is publishing to path...."  in the add on log for sometimes one camera or more.  If I start and stop the add on they start working again for a while.  Does that sounds like an add on issue?

**Reply by mrlt8** [reply]
Not related to HA. It's sort of a known problem that I've been trying to figure out since it should restart on its own.

Something that you could do is check your 2.4ghz network and ensure good coverage and channel availability for the cameras.

**Reply by chubRock1** [reply]
Ah, okay.  I do have one cam that reports 44% signal, but the others are at around 80% or higher and it can be random which ones drop off after the time.   Happy to help with troubleshooting if there is anything I can do.

**Reply by mrlt8** [reply]
I just pushed v0.6.4 that *might* help.
I believe the cams will sometimes change the resolution of the stream which could cause the bridge to get stuck. The new update will try to detect and restart the streams if the resolution doesn't match for more than 30 frames.

Please let me know if this helps!

**Reply by chubRock1** [reply]
Thanks for the update, I do see some attempts in the log to restart ffmpeg, but I am not seeing any real change in the behavior.  The streams still end up showing as no one is publishing to path...  let me know if sending you any logs would be helpful.

**Reply by unrly** [reply]
Having similar experiences, it can vary from a few hours to several hours and I thought it was tied to one camera (both of my V2s are reporting 75%+ wifi) but the one that was more reliable dropped this morning after the update.

I restarted the addon a little bit ago, and it seems more stable this time around. Also changing my 2.4ghz channel to see if that helps at all since I found a lot of interference last night. How can I enable debug logs in HA? I tried with the 'unused options' but it didn't seem to stick. Would just be curious to see/can also help in any way.

**Comment by mrlt8** [comment]
There have been several mentions of setting not sticking.
Could a Home Assistant expert have a look at the [config.json](https://github.com/mrlt8/docker-wyze-bridge/blob/main/app/config.json) to see if everything looks ok?

**Reply by Taylorp913** [reply]
I had first hand experience with the config not sticking in the supervisor web ui and I have not seen that before. I cannot tell if anything is wrong in the config.json, but I might try adding the same keys from the schema object to the options object. Not sure if that is nessesary for optional parameters, but here we are just guessing. I'll let you know if find anything else.

**Reply by Taylorp913** [reply]
Reading online, users will usually see issues saving the add-on config if there is a error is the config, but it does a very poor job of reporting to the user what caused it. You are using the schemas which should prevent issues like that though.

**Reply by mrlt8** [reply]
I tried setting the optional keys under "options", but then they fail validation when saving in the gui because of the null/false values: _Failed to save add-on configuration, not a valid value._

Another weird thing is that it will randomly add a blank value for one of the keys even though those fields were never touched (QUALITY and FFMPEG_CMD in this example):
```yaml
WYZE_EMAIL: email@email.com
WYZE_PASSWORD: password
API_THUMB: true
LAN_ONLY: true
QUALITY: ''
IGNORE_OFFLINE: true
FFMPEG_CMD: ''
FRESH_DATA: true
```

**Reply by mrlt8** [reply]
I thought it might have had something to do with using "legacy" mode, so I ended up going down a rabbit hole of [reconfiguring](https://github.com/mrlt8/docker-wyze-bridge/pull/123) the container to use HA's options.json so we could disable legacy mode, and the stickiness seems like it might have gone away, but I'm still getting the phantom values being set:

```yaml
RTSP_THUMB: false
FILTER_MACS: ''
IGNORE_OFFLINE: true
```

**Comment by tbrausch** [comment]
Having issues trying to get the Wyze Doorbell Cam working with this addon.  It recognizes the doorbell, but I get the following repeated over and over in the Docker Wyze Bridge log:

```
2021/09/15 07:31:10 [Doorbell] 🎉 Starting 1080p 120kb/s Stream for WyzeCam DOORBELL in "LAN mode" FW: 4.25.0.245 IP: 192.168.1.14 WiFi: 81%
2021/09/15 07:31:10 [Doorbell] [FFMPEG] Wrong resolution: 4
2021/09/15 07:31:10 [Doorbell] 🧹 Cleaning up FFMPEG...
2021/09/15 07:31:12 [Doorbell] 🎉 Starting 1080p 120kb/s Stream for WyzeCam DOORBELL in "LAN mode" FW: 4.25.0.245 IP: 192.168.1.14 WiFi: 79%
2021/09/15 07:31:12 [Doorbell] WARNING: Lost frame
2021/09/15 07:31:12 [Doorbell] [FFMPEG] Wrong resolution: 4
2021/09/15 07:31:12 [Doorbell] 🧹 Cleaning up FFMPEG...
2021/09/15 07:31:14 [Doorbell] 🎉 Starting 1080p 120kb/s Stream for WyzeCam DOORBELL in "LAN mode" FW: 4.25.0.245 IP: 192.168.1.14 WiFi: 81%
```
I have seen references to fixes for the doorbell cam in the change log, so am I correct in assuming this should work?  I get no streams in RTSP, RTMP, or HLS.  All other cameras work, including V2 and V3 cams.  Thanks in advance for any help.

**Reply by mrlt8** [reply]
Sorry about that! There's been a whole mess due the frame issue on the WYZEDB3, and I've been waiting to hear back from some users on another thread as I don't have a doorbell to test thing out on and wanted to confirm the issue before pushing out the next version.

I've pushed a potential patch to the main branch. Would appreciate it if you could test it out before I tag it as a release.
If you're pulling the image, the changes should be available on `mrlt8/wyze-bridge:edge`

**Comment by tbrausch** [comment]
I would very much like to help out, but I am running Wyze Bridge as a Supervisor Addon.  Is there still a way for me to test this out?  I downloaded the zip file for the edge development path but I'm not sure how to install it.

**Reply by mrlt8** [reply]
I went ahead and tagged v0.6.6 as a release, so you should be able to upgrade in HA. Please let me know if this helps.

**Comment by tbrausch** [comment]
Afraid I am getting the same results:
```
2021/09/16 03:30:34 [Doorbell] 🎉 Starting 1080p 120kb/s Stream for WyzeCam DOORBELL in "LAN mode" FW: 4.25.0.245 IP: 192.168.1.14 WiFi: 79%
2021/09/16 03:30:34 [Doorbell] [FFMPEG] Wrong resolution: 4
2021/09/16 03:30:34 [Doorbell] 🧹 Cleaning up FFMPEG...
2021/09/16 03:30:36 [Doorbell] 🎉 Starting 1080p 120kb/s Stream for WyzeCam DOORBELL in "LAN mode" FW: 4.25.0.245 IP: 192.168.1.14 WiFi: 80%
2021/09/16 03:30:36 [Doorbell] [FFMPEG] Wrong resolution: 4
2021/09/16 03:30:36 [Doorbell] 🧹 Cleaning up FFMPEG...
```

**Reply by mrlt8** [reply]
hmm, looks like your stream is reverting to [SD](https://github.com/kroo/wyzecam/blob/a5ee93a5a5c396e7f2bac58220f17eafe524fb18/wyzecam/tutk/tutk.py#L59-L69) or is taking a long time to switch to HD..?

A couple of things to try in Supervisor > Wyze bridge  > Configuration > Edit in YAML:

Set `IGNORE_RES: true`  (may need to stop the add-on and hit save a couple of times till the options save). 
If this works, try opening the :8554 RTSP stream in a media player like VLC and verify if the stream is 480 x 640 or 1296 x 1728 or switches from one to the other. 

If it does switch from SD to HD, then we'll need to figure out how long it takes  by removing the previous `IGNORE_RES` and adding `BAD_FRAMES: 50` and keep increasing the number by 25 _if_ you see the "Wrong resolution" error.

**Reply by chubRock1** [reply]
Just chiming in to say that I am not seeing a change to the no one is publishing to path error I get after some running time using 0.6.6

**Comment by tbrausch** [comment]
So, let me see if I can explain what I am seeing.  I changed the configuration to include IGNORE_RES: true and restarted the addon.  I was able to display the live doorbell camera feed in a picture entity card, with the video continually changing to "all green" for an instant about every 15 or 20 seconds.  I should note that the video was rotated CCW by 90 degrees. 

Displaying the camera in VLC:
RTSP: Trying to display the camera in VLC via rtsp caused an error - unable to display stream. It worked for the other Wyze cameras.
RTMP: Displaying the rtmp stream in VLC displayed what could be called a pixelated mess, but it did display - sort of like a kaleidoscope.  This was true for all of the other Wyze cameras, also.  

Displaying the cameras via HHS resulted in the same pixelated screen for all cameras.

I also use Frigate with the WebRTC option.  Displaying all the Wyze cams in a WebRTC card caused the same pixelated display (although this worked for these cameras previously).

I then removed IGNORE_RES: true and added BAD_FRAMES: 50.  This ended up generating many errors in the logs and the cameras did not work:

```
2021/09/16 17:13:41 [Porch2] 🧹 Cleaning up FFMPEG...
2021/09/16 17:13:41 [Doorbell] [FFMPEG] '>=' not supported between instances of 'int' and 'str'
2021/09/16 17:13:41 [Doorbell] 🧹 Cleaning up FFMPEG...
2021/09/16 17:13:41 [Patio] 🎉 Starting 1080p 120kb/s Stream for WyzeCam V3 in "LAN mode" FW: 4.36.3.19 IP: 192.168.1.52 WiFi: 82%
2021/09/16 17:13:41 [Patio] [FFMPEG] '>=' not supported between instances of 'int' and 'str'
2021/09/16 17:13:41 [Patio] 🧹 Cleaning up FFMPEG...
```

**Reply by mrlt8** [reply]
Thanks for the info!  I'm still looking into this.
Just to confirm, was the doorbell working prior to v0.6.x? and any firmware changes since?

**Comment by tbrausch** [comment]
I have four V3 cameras and 2 V2's that, until I found your addon a couple weeks ago, I was considering trashing.  It's been almost a year since Wyze promised RTSP on the V3s and still nothing.  But your implementation is fantastic - the Wyze Cams turned from gargage into sweet little gems.  They are now faster, more reliable, and with frigate installed become the perfect camera solution.  To answer your question, the doorbell has never worked, but I just bought it last week.  Let me know if there is any other way I can help.
